### ISSUE TITLE
EMA difficulty algorithm

### ISSUE THREAD

**Issue by zawy12** [zawy12/difficulty-algorithms#4]
```
# Jacob Eliosoff's EMA (exponential moving average)
# https://en.wikipedia.org/wiki/Moving_average#Application_to_measuring_computer_performance
# ST = solvetime, T=target solvetime
# height = most recently solved block
# N=60
# There's a "2" in the exponent below to make the above N comparable to SMA, WT-144, & WWHM
# MTP should not be used

for (i=height - 10 to height) {  # check timestamps starting 10 blocks into past)
   previous_max = max
   max=timestamp[i] if timestamp[i] > max
}
ST = max - previous_max
ST = max(T/100,min(T*10, ST))
next_D = previous_D * ( T/ST + e^(-ST*2/T/N) * (1-T/ST) )
```

**Comment by h4x3rotab** [comment]
Still confused by the following code:
```
previous_max=0,  max=0
for (i=height - 10 to i <= height ) {  
   if (previous_max < timestamp[i] ) { max = timestamp[i]  }
   ST = max - previous_max
   previous_max = max
}
```

1) Why do you choose the last 10 blocks to find the 1st and 2nd latest block time? I thought 3-4 blocks are fine as it's hard to manipulate all these blocks.
2) If I understand it correctly, the `ST = max - previous_max` and `previous_max = max` should be inside the `if` condition because otherwise `previous_max` will become `max` whenever the max is updated or not. Am I wrong?

```
previous_max=0,  max=0
for (i=height - 10 to i <= height ) {  
   if (max < timestamp[i] ) {
      max = timestamp[i] 
      ST = max - previous_max
      previous_max = max
   }
}
```

**Comment by zawy12** [comment]
Crap.  I edited it above so maybe now it is right.

Concerning 10: I have seen 6 blocks in a row that tried to be older than MTP.  Although a subsequent MTP may not be older than a previous MTP, maybe all coins do not follow that rule.  I mean maybe there is a coin that could have -1000 from the previous timestamp 6 times in a row.  We can't simply make them =0 because it would be subtracted from an honest timestamp after it which would be a large solvetime.  If an attacker did a "negative" every other bock, difficulty could go to zero.

**Comment by zawy12** [comment]
Your method is not what I'm trying to do because if the most recent block is negative, then the ST is equal to the solvetime of the previous block.  

The problem is that I've done a bad job of copying Tom Harding's method that's used in the WHM.  It seems to not easily convert over to the EMA.  
```
for ( i=height-10 to i<=height )  {
    if (timestamp[i] <= prev_max ) {  ST = 0 }
    else { 
           ST = timestamp[i] } - prev_max
           prev_max = timestamp[i]
   }
}
```

**Comment by zawy12** [comment]
Testing a sequence of blocks with T = ST but assigned timestamps:
``` 1, 2, 3, 4, -3, 6, 7, -5, 9, 10, ```
Gives STs:
```1,1,1,0,2,1,0,2,1```
OK, I think that's the behavior I want.

**Comment by zawy12** [comment]
Another testing of timestamps  (again real ST for each is magically 1, and the guys out of sequence are the liars).  Only 1, 2, 3 and last 15, 16, and 17 are correct
```1,2,3,10,2,12,2,9,7,9,7,6,15,14,15,16,17```
STs desired and should be given by above code:
```1,1,7,0,2,0,0,0,0,0,0,3,0,0,1,1```
There were 16 ST's and their average is again 1 like we wanted.

**Comment by h4x3rotab** [comment]
Thanks for the correction.

Another quest: by assigning 0 to ST, ST will always be 0 when the last block is lying (consider 1,2,3,4,5,6,7,8,9,1), though it will be clipped to T/200 later. In this case, ST is no longer the time between the 1st and 2nd latest block. Is that intended behavior?

**Comment by zawy12** [comment]
Yes.  You made me realize I could simplify the code which I did.  We just get the max time from previous blocks (i<height instead of i<=height in the loop) and subtract that from the most recent block after the loop.

**Comment by zawy12** [comment]
BTW here is the problem with simple ST=0 if ST<0 if a miner has 33% hashpower:
``` 
timestamps:
2,3,4-6,5,6,7-6,8,9,10-6,11,12,1-6,14
STs:
1,0,7,1,0,7,1,07
```
Average ST = 2.67 so D = correct_D/2.67 at the end of N blocks, and about D/5 in 2N.  It keeps driving difficulty down until everyone is solving in 0 time, but the timestamps are still bad:
```
0,0,0-6,0,0,0-6
gives solvetimes
0,0,6,0,0,6
```
So difficulty will still keeping dropping by D/2 every N blocks.

**Comment by h4x3rotab** [comment]
Thank you for clarification. I’m working on the implementation for BTG
testnet :)

**Comment by zawy12** [comment]
I found another timestamp cheat. Be sure to use new code on first post above. Your limit value should be 7

**Comment by h4x3rotab** [comment]
Just came up with another detailed question:

```
maxT=timestamp[height-limit-1]
for ( i = height - limit to i < height )  { maxT = max(maxT, timestamp[i]) }
```

If `height` is the height of the last solved block, the last a few blocks should be in the range [height-limit+1, height]. But the code here is [height-limit-1, height-1].

**Comment by zawy12** [comment]
Yes, that is my intention. The loop just finds the previous max.

**Comment by h4x3rotab** [comment]
Ah, just got it. But shouldn't it be [height-limit+1, height-1]?

**Comment by zawy12** [comment]
if height = previous solved block then what I have is correct.  If height = current block being solved, then what you have is correct.

For simplicity, maybe you should use [WHM algorithm](https://github.com/zawy12/difficulty-algorithms/issues/3) instead.  Masari is having excellent success with it.

**Comment by h4x3rotab** [comment]
Just finished the first implementation. The constants and the details may need to be further adjusted. The implementation is based on fixed point calculation and use Tyler series to calculate exp(). I've checked the precision and it looks perfect. However I didn't check the offset problem. Would you like to take a look?

https://github.com/h4x3rotab/BTCGPU/commit/1f681b47e0685f2bd3c1712bbc1060d297493476

**Comment by zawy12** [comment]
Just to confirm things: 
The following timestamp must be at nHeight.
 pindexLast->GetBlockTime()


The T/200 could be set to 2 or 3, and probably 1 is fine.
solve_time = std::max((uint64_t)(T / 200), std::min((uint64_t)(T * limit), solve_time));  // 200???? 3?

You could set the following to limit=7 and it will work for all N > 50.  But limit =6 is safer for N<60.  The problem is if someone assigns timestamp 7xT into future, difficulty will drop about 7/50 = 14%.  With 6 it will be 12%.  
 int limit = std::min(7, std::max((N * 100 / 89) - N, 10));

I can't tell if your series expansion is correct, but I know the following works perfectly:
e^x = 1+x+x^2/2
I expanded this to integer math here:https://github.com/zawy12/difficulty-algorithms/issues/17

k= ST*T*M  // M=25 instead N=50next_target = prev_target / 2 * ( 2*(T*M)^2-2*T^2*M+k*(k+2*(T*M)^2-T^2*M) ) / (T*M)^2 
The LWMA (WHM) is a little bit better than EMA and uses the same timestamp handling if you can't allow negative solvetimes.  Two monero coins are now using it with great success.
https://github.com/zawy12/difficulty-algorithms/issues/3

**Comment by h4x3rotab** [comment]
T/200: Opps! Mistake caught.
Series expansion: Compared with expl. If negative ST is allowed, it can be tweaked a little bit to work.

It has been a few weeks since I paid close attention to the algorithm experiments. This is just a proof-of-concept about implementing float operators by fixed point. As you mentioned, LWMA is actually better than this.

**Comment by zawy12** [comment]
BTC candy tells me this morning your LWMA is not working good for them on their live coin.  I am investigating it now.

**Comment by h4x3rotab** [comment]
Thank you! I will send you a dump of testnet timestamps and difficulties
tomorrow.

**Comment by zawy12** [comment]
I finally determined that nothing was wrong with their LWMA.....they just have really bad mining situation
They said it is doing better than digishield was doing.




    On Thursday, May 3, 2018, 3:39:46 PM EDT, h4x3rotab <notifications@github.com> wrote:  
 
 Thank you! I will send you a dump of testnet timestamps and difficulties
tomorrow.
